### ISSUE TITLE
Embed strings without the string quotes

### ISSUE THREAD

**Issue by sangwinc** [maths/moodle-qtype_stack#265]
This is related to issue #260.

We have a proposal to change the way STACK deals with CAS expressions which are only a string.  In particular, when @p@ is a pure string we should not embed it within a maths environment and an \mbox.  The "value" field (in the abacus branch) will keep the quote marks, so this feature is not satisfied by that.

Comments on the prototype code welcome.

**Comment by aharjula** [comment]
Merge with abacus... And make a new block with shorthand syntax to do this e.g. {%...%} in addition to the existing {@...@} and {#...#}. You should be able to test creation of such a block yourself should not require much to copy the current {#...#} and strip the quotes and unescape the string there before output.

As we talked in Newcastle this would be the ideal way for you to familiarise yourself with the Abacus parser and block system.

And no I don't think {%...%} is the correct one % is a bad choice but having those outer matching {}-wrappers is something I wish it to have.

**Comment by aharjula** [comment]
c28f8ee is better than having a separate {%...%} or such, less syntax and less code. So lets go for special handling of values that are just strings, it should case no trouble for people that that edge case behaves differently. Just ensure that the string is unescaped correctly, i.e. quotes within the string and so on.

**Comment by aharjula** [comment]
Now we just need to translate that c28f8ee to the [latex-block](https://github.com/maths/moodle-qtype_stack/blob/abacus/stack/cas/castext/latex.class.php).

**Comment by sangwinc** [comment]
Ok, there is no rush here.  Let's give this at least 24 hours to think about and see any other comments...

**Comment by sangwinc** [comment]
Thanks Matti,
This is now merged into master.
Chris

**Comment by aharjula** [comment]
Did you check that the escapes are being correctly handled? I.e. can one generate a string containing LaTeX and have the slashes come out correctly? What happens with a value of `"\\sqrt{5}"` Does it render correctly as `\sqrt{5}` or incorrectly as `\\sqrt{5}`? Remember that the utils library has a function for handling that unescaping should it be needed...

**Comment by aharjula** [comment]
There really should be a unit test to confirm that one can get correctly unescaped content out.

**Comment by sangwinc** [comment]
Matti, 
I think the current code does render this "correctly" as you suggest.  I've added another unit test to confirm this.  Thank you for suggesting this.  Please feel free to suggest more test cases as needed.

**Comment by sangwinc** [comment]
Matti,

Looking at the code in your new iss265 branch, these changes appear to break valid maxima syntax such as p:"+"(1,2,3).  When using + as a prefix operator you need to give the name as a string!

Chris

**Comment by aharjula** [comment]
Indeed which is why I wished you to comment on that new `check_string_usage()`. Just add `(` here [here](https://github.com/maths/moodle-qtype_stack/blob/iss265/stack/cas/casstring.class.php#L683) and that is again valid. I probably applied rules based on experience with other languages, so that went over but I assume other checks in that function are valid?

**Comment by sangwinc** [comment]
Thanks Matti,

I've added this in as you suggested, and am running the unit tests. 

I've run all the question tests on the "Demonstration questions" site.  Good news is that nothing was flagged up.  Is is possible to run the question tests on the whole abacus project with this code?

**Comment by aharjula** [comment]
Eventually, Chris, but as said there are very few tests there... Hopefully I get some time next week to do a simple database dump of all keyvals from Abacus and then run this on them. But for now I am waiting for comments on how that re-factor looks to others, style and so on and specifically comments from Tim about the possible issues he might see with the trade-offs, between speed and memory, that I made although I don't think this uses any more memory, might actually use less due to less GC.

**Comment by sangwinc** [comment]
Matti,
I hope this (https://github.com/maths/moodle-qtype_stack/commit/70514646cae6561fcf1bdddc01e276f4564eadeb) is what you had in mind here.  I've run all the unit tests on this new code.  Nothing new fails, but we still have a few answer tests related to missing brackets which I'd like to look into.

**Comment by sangwinc** [comment]
Matti,
In merging this code with equiv_input this test seems to fail:  https://github.com/maths/moodle-qtype_stack/blob/iss265/tests/castext_test.php#L572
Was that intentional?  Would you like to make "Hello world"+x^2 an invalid statement?

**Comment by aharjula** [comment]
The idea behind the latter part of `check_string_usage()` was to make certain obviously invalid inputs with "strings" invalid in the validation, but it seems unlikely that we could really identify what cases are really invalid in a static analysis that is limited to a context of a single casstring. So probably we should drop [all this](https://github.com/maths/moodle-qtype_stack/blob/iss265/stack/cas/casstring.class.php#L657-L713), what I wanted to do is not really possible there so lets just roll back on that new test.

What I really wanted is something that really checks the types, something like [Tern](http://ternjs.net/doc/demo/index.html#simple), but it makes more sense to do that kind of checking on the editor side. Lets just say that once I get to implementing the stateful question type it will have an editor doing that kind of type validation.

**Comment by sangwinc** [comment]
Thanks Matti,
I've been talking with Christian about more flexible parsers written in javascript, which I hope we can work on in the near future.  That might be a good place to fix this and reinstate the checks you want.

**Comment by sangwinc** [comment]
I've merged your suggested changes into the equiv_input branch to ensure this is tested in parallel with the ongoing developments there.  

I've removed all your early bail-outs in the cassting classes.  This is needed because even when we already know a casstring is invalid we need to (1) give feedback to users, and (2) split up the key-vals pairs for subsequent use.  The order of the test also matters, as some of the tests for validation have side effects on the casstring value.
